Refactor Norm to allow easier extensions

igorcadelima opened the pull request:
The following changes were made to allow easier creation of classes with the same interface as the Norm class:

- Remove NormFactory due to no real need for it and unsuitability if aiming to allow extensions of Norm
- Add INorm interface with essential methods for norms
- Add AbstractNorm, which implements INorm, with base functionality of Norm
- Refactor Norm and make it an extension of AbstractNorm
- Replace declarations/returns of Norm with INorm

jomifred commented:
Thanks Igor, norm factory is used by the parser and allows the programmer to set another norm factory. Is this feature kept in your changes?

igorcadelima commented:
Hi @jomifred !

Unfortunately, the factory only makes sense when the norm defined by the programmer also needs the same parameters defined in the factory method in order to be instantiated. I can tell one example of that myself based on my masters project where we need to have more parameters in order to instantiate a norm.

This is actually a drawback of the factory design pattern. If you need to have more parameters in the factory method then NormFactory::createNorm would have to be changed as well as its implementations for every change. 

Answering your question, no, it's not. I removed it in 221d013d7da5bac249e271dd278f4d73fbf9217f. As you can see, I changed the parser as well to reflect this change.

Please let me know if you've got more questions.

jomifred commented:
Yes, I know that limitation of the factory. However another project that uses NPL needs that the parser uses that project's norm implementation instead of the class Norm.... The current implementation (although not ideal as you pointed) allows that (without changing the parser). 

In cases where more parameters are required, the programmer can extend the parser and call his own factory. 

So, I do not see the necessity of removing the factory.

(all other changes -- INorm, Abstract Norm, ... are fine)

igorcadelima commented:
Just to make sure, in cases where more parameters are required, you mean that the programmer should:

1. Create a class that implements NormFactory (let's call it NormFactoryExtension)
2. Create a class that extends Norm (let's call it NormExtension)
3. In NormFactoryExtension, override NormFactory::createNorm
4. In NormFactoryExtension, **overload** createNorm to make it return an instance of NormExtension
5. In NormExtension, override Norm::getFactory to make it return an instance of NormFactoryExtension
6. Extend the parser to make it use NormExtension::getFactory

Is that right?

By _another project_ you mean another existing project or any other project? If the former, yeah... that'd definitely be an issue.

igorcadelima commented:
I see your point now, @jomifred ! Sorry! I'll let you know when I restore the parser, NormFactory, and Norm::getFactory :)

jomifred commented:
Steps 1...4 are fine

5. Instantiate the parser and use setNormFactory(<your instance of the factory>)

no need of step 6.

igorcadelima commented:
Hi @jomifred !

I just reverted the commit as you asked. I guess it's all good now :)

----

I believe that step 5 would be necessary in order to really use NormFactoryExtension. Otherwise an instance of the old NormFactory would be returned when calling NormExtension::getFactory.

I also believe that step 6 would be necessary because the parser creates a norm using 
```
nFac.createNorm(id.image,h,(LogicalFormula)b); 
```
Calling setNormFactory wouldn't be enough as there'd be no passing of new arguments.

Cheers.

jomifred commented:
Thanks Igor, I merged your commits.
